# Занятие 7. Поиск отдельных видео, хорошо репрезентирующих совокупность всех релевантных видео


[x] Определите тип шкалы переменных statistics.viewCount, statistics.likeCount, statistics.dislikeCount, statistics.commentCount.
[] Определите, какие методы описательной статистики релевантны исследуемым переменным с т.з. типа шкалы последних.
[] Примените выбранные методы описательной статистики относительно каждой исследуемой переменной, чтобы познакомиться с ней: (а) постройте её распределение,
(б) найдите значение, репрезентирующее по данной переменной всю совокупность видео,
(в) оцените, насколько этому значению можно доверять,
(г) найдите видео, наиболее репрезентирующие совокупность по исследуемым переменным; посмотрите эти видео и комментарии к ним; проведите анализ в качественной парадигме.

[] Определите тип шкалы переменной categoryName.
[] Определите, какие методы описательной статистики релевантны исследуемой переменной с т.з. типа шкалы последней.
[] Примените выбранные методы описательной статистики относительно исследуемой переменной, чтобы познакомиться с ней: (а) постройте её распределение,
(б) найдите значение, репрезентирующее по данной переменной всю совокупность видео,
(в) оцените, насколько этому значению можно доверять,
(г) найдите видео, наиболее репрезентирующее совокупность по исследуемой переменной; посмотрите это видео и комментарии к нему; проведите анализ в качественной парадигме.
[] Выводы впишите в скрипт в формате комментариев или Markdown.
[] Обогатите визуализацию описательной статистики для каналов и видео. выгруженных на занятиях соответствующей уч.дисциплины в социологическом бакалавриате ВШЭ, посредством построения интерактивных диаграм.
[] Постройте облако слов на основе любой из выгруженных на прошлых занятиях текстовых характеристик
[] Если всё получилось, нажмите Kernel -> Restart & Run All для автоматического последовательного запуска всех чанков.
[] Покажите результат преподавателю.

In [1]:
# 1 загрузка конфигов и путей
import pathlib

import numpy as np
import yaml

ROOT = pathlib.Path().resolve().parent.parent

with open(ROOT / "config.yml", "r") as f:
    config = yaml.safe_load(f)

In [2]:
# 2 Загрузка данных о видео

DATA_PATH = ROOT / config['api']['DATA_PATH']
print(DATA_PATH)

/Users/andrey/PycharmProjects/DataMiningProjects/data


In [3]:
# 3 импорт необходимых библиотек

import randan as rd
import googleapiclient.discovery as api
import pandas as pd
import numpy as np

In [4]:
# 4 загрузка данных

df_korean = pd.read_excel(DATA_PATH / 'video' / 'Video_characteristics_Корейская_косметика.xlsx')

In [5]:
# 5 загрузка данных

df_korean.head()

kind                         etag           id  \
0  youtube#video  hFWHeWO-R7oU4c2XL_XGUFSW0Ro  Wi49rKRM85U   
1  youtube#video  dH5eRX5ktvZc5h6hZepKAsK6zSo  1ErvLvQoYxc   
2  youtube#video  VAaYvq0w2V6n-a6E6upg2A1iJbE  aiY0ymL25Pw   
3  youtube#video  Xu0qxfp0-q2X2DrGsPlfAAn6nqM  JA9E4TCVsDU   
4  youtube#video  1G4hAX30sQ9cKXA2lX2UD6wUbuY  NAeXVpMTiT4   

    snippet.publishedAt         snippet.channelId  \
0  2022-06-20T12:57:48Z  UCphnZS52v_lkv5RSl5n70tQ   
1  2019-03-21T10:23:49Z  UC6C0oFYlpKkFGG-ykqM4fYw   
2  2021-08-20T08:51:18Z  UCiT7bPorr4rKJU8WFBZA9Og   
3  2015-12-06T14:30:40Z  UCNazbf-kLY6E8TO_o48ZExQ   
4  2018-04-20T15:19:56Z  UCmXGqfEsluA1fID55bVexjw   

                                       snippet.title  \
0  한국인들은 절대 안사는 화장품 КОСМЕТИКА КОТОРАЯ КОРЕЯНКИ ВО...   
1  ТЫ ДОЛЖНА ЭТО ПОПРОБОВАТЬ! ЛУЧШАЯ КОРЕЙСКАЯ КО...   
2  РЕАКЦИЯ дерматолога: 10 этапов корейского уход...   
3  КАК Я ИСПОРТИЛА КОЖУ: КОРЕЙСКАЯ КОСМЕТИКА\Tony...   
4  🇰🇷 НЕВЕРОЯТНАЯ КОРЕЙСКАЯ КОСМЕТИКА! Топовые шт...   

                                 snippet.description  \
0  ⚫подписаться: https://goo.gl/3QbeUf\n---------...   
1  Хорошие мои, сегодня я расскажу вам про свой к...   
2  Мой телеграм:\nhttps://t.me/dr_amina_pirmanova...   
3  Привет! Корейская косметика снова набирает обо...   
4  Магазин корейской косметики https://goo.gl/UJG...   

                   snippet.thumbnails.default.url  \
0  https://i.ytimg.com/vi/Wi49rKRM85U/default.jpg   
1  https://i.ytimg.com/vi/1ErvLvQoYxc/default.jpg   
2  https://i.ytimg.com/vi/aiY0ymL25Pw/default.jpg   
3  https://i.ytimg.com/vi/JA9E4TCVsDU/default.jpg   
4  https://i.ytimg.com/vi/NAeXVpMTiT4/default.jpg   

   snippet.thumbnails.default.width  snippet.thumbnails.default.height  ...  \
0                               120                                 90  ...   
1                               120                                 90  ...   
2                               120                                 90  ...   
3                               120                                 90  ...   
4                               120                                 90  ...   

  statistics.favoriteCount  statistics.commentCount  \
0                        0                   1065.0   
1                        0                    378.0   
2                        0                   1777.0   
3                        0                    324.0   
4                        0                    687.0   

                                    player.embedHtml  \
0  <iframe width="480" height="270" src="//www.yo...   
1  <iframe width="480" height="270" src="//www.yo...   
2  <iframe width="480" height="270" src="//www.yo...   
3  <iframe width="480" height="270" src="//www.yo...   
4  <iframe width="480" height="270" src="//www.yo...   

                        topicDetails.topicCategories  snippet.defaultLanguage  \
0  ['https://en.wikipedia.org/wiki/Lifestyle_(soc...                      NaN   
1  ['https://en.wikipedia.org/wiki/Lifestyle_(soc...                      NaN   
2  ['https://en.wikipedia.org/wiki/Health', 'http...                      NaN   
3  ['https://en.wikipedia.org/wiki/Lifestyle_(soc...                      NaN   
4  ['https://en.wikipedia.org/wiki/Lifestyle_(soc...                      NaN   

   liveStreamingDetails.actualStartTime liveStreamingDetails.actualEndTime  \
0                                   NaN                                NaN   
1                                   NaN                                NaN   
2                                   NaN                                NaN   
3                                   NaN                                NaN   
4                                   NaN                                NaN   

   liveStreamingDetails.scheduledStartTime  statistics.dislikesCount  \
0                                      NaN                     595.0   
1                                      NaN                     192.0   
2           

In [6]:
# 6 

df_korean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1315 entries, 0 to 1314
Data columns (total 41 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   kind                                     1315 non-null   object 
 1   etag                                     1315 non-null   object 
 2   id                                       1315 non-null   object 
 3   snippet.publishedAt                      1315 non-null   object 
 4   snippet.channelId                        1315 non-null   object 
 5   snippet.title                            1315 non-null   object 
 6   snippet.description                      1045 non-null   object 
 7   snippet.thumbnails.default.url           1315 non-null   object 
 8   snippet.thumbnails.default.width         1315 non-null   int64  
 9   snippet.thumbnails.default.height        1315 non-null   int64  
 10  snippet.thumbnails.medium.url            1315 no

In [7]:
from randan.descriptive_statistics import ScaleStatistics, OrdinalStatistics, NominalStatistics

# Изучение интервальных переменных

** Переменные `viewCount`, `likeCount`, `dislikesCount`, `commentCount` относятся к интервальному типу**

In [8]:
# 7 описательная статистика



ss = ScaleStatistics(df_korean, ['statistics.viewCount',
                          'statistics.likeCount',
                          'statistics.commentCount',
                                 'statistics.dislikesCount'],
                     normality_test=True)



SCALE STATISTICS SUMMARY
------------------



/Users/andrey/PycharmProjects/DataMiningProjects/.venv/lib/python3.10/site-packages/randan/descriptive_statistics.py:491: FutureWarning: this method is deprecated in favour of `Styler.format(precision=..)`
  .set_precision(n_decimals))



NORMALITY TESTS
------------------



/Users/andrey/PycharmProjects/DataMiningProjects/.venv/lib/python3.10/site-packages/randan/descriptive_statistics.py:460: FutureWarning: this method is deprecated in favour of `Styler.format(precision=..)`
  .set_precision(n_decimals))


In [9]:
# 8 Суммарные значения
ss.summary()

N  mode  median          mean    25%      75%  \
statistics.viewCount      1315.0  48.0  1357.0  22153.692776  249.5  7802.50   
statistics.likeCount      1284.0   1.0    42.5    972.755452    8.0   305.25   
statistics.commentCount   1299.0   0.0     4.0     68.354119    0.0    36.50   
statistics.dislikesCount  1313.0   0.0     0.0     19.728104    0.0     5.00   

                          interquart. range  interquart. range (norm.)  min  \
statistics.viewCount                7553.00                   0.004144  0.0   
statistics.likeCount                 297.25                   0.003632  0.0   
statistics.commentCount               36.50                   0.002803  0.0   
statistics.dislikesCount               5.00                   0.001067  0.0   

                                max      range            std           var  \
statistics.viewCount      1822564.0  1822564.0  100170.569684  1.003414e+10   
statistics.likeCount        81849.0    81849.0    4826.984719  2.329978e+07   
statistics.commentCount     13022.0    13022.0     431.534288  1.862218e+05   
statistics.dislikesCount     4684.0     4684.0     157.367981  2.476468e+04   

                          entropy coef.  quality var.  
statistics.viewCount           0.985176      0.999694  
statistics.likeCount           0.863502      0.990473  
statistics.commentCount        0.683720      0.899213  
statistics.dislikesCount       0.438101      0.605393


В качестве используемой меры центральной тенденции мы предлагаем использовать медиану.

Общее количество данных - 680


 ### Переменная `viewCount`

Median - 6725

 ### Переменная `likeCount`

Median - 42.5

 ### Переменная `likeCount`

Median - 42.5



 ### Переменная `commentCount`

Median - 4


In [10]:
# 9 вспомогательная функция

def get_typical_by_median(df, column):

    lower = np.nanpercentile(df_korean[column], 49)
    upper = np.nanpercentile(df_korean[column], 51)
    tmp_df = df[(df[column] >= lower) & (df[column] <= upper)][
    ['id', 'snippet.channelId','snippet.title','snippet.description', 'snippet.publishedAt',
        'snippet.tags', 'snippet.categoryId', 'categoryName']]

    display(tmp_df)
    return tmp_df


## statistics.viewCount

In [ ]:
# 10 График распределения statistics.viewCount
import plotly.express as px

fig = px.histogram(df_korean, y="statistics.viewCount",
                   marginal="box")
fig.show()

In [ ]:
# 11 Типичные видео statistics.viewCount 2%

view_df = get_typical_by_median(df_korean, 'statistics.viewCount')

## statistics.likeCount

In [ ]:
# 12 График распределения statistics.likeCount

import plotly.express as px

fig = px.histogram(df_korean, y="statistics.likeCount",
                   marginal="box")
fig.show()

In [ ]:
# 13 Типичные видео 2%

likes_df = get_typical_by_median(df_korean, 'statistics.likeCount')

## statistics.dislikesCount

In [ ]:
# 14 График распределения statistics.dislikesCount

import plotly.express as px

fig = px.histogram(df_korean, y="statistics.dislikesCount",
                   marginal="box")
fig.show()

In [ ]:
# 15 Типичные видео 2%

dislike_df = get_typical_by_median(df_korean, 'statistics.dislikesCount')

## statistics.commentCount

In [ ]:
# 16 График распределения statistics.dislikesCount

import plotly.express as px

fig = px.histogram(df_korean, y="statistics.commentCount",
                   marginal="box")
fig.show()

In [ ]:
# 17 Типичные видео 2%

comment_df = get_typical_by_median(df_korean, 'statistics.commentCount')

In [ ]:
# 18 Intersection of 2 %
id_set = set()

interval_sets = {'View': set(view_df['id']),
                 'Like': set(likes_df['id']),
                 'Dislike': set(dislike_df['id']),
                 'Comment': set(comment_df['id'])}




In [ ]:
# 19 Комбинации
from itertools import combinations

comb_2 = combinations(interval_sets.keys(), 2)
comb_3 = combinations(interval_sets.keys(), 3)

In [ ]:
# 20 Пересечения общих частей по 2-м совпадениям

for pair in comb_2:
    print(f'Общая часть от пересечения {pair[0]} + {pair[1]}', len(interval_sets[pair[0]] & interval_sets[pair[1]]))

In [ ]:
# 21 Пересечения общих частей по 3-м совпадениям

for pair in comb_3:
    print(f'Общая часть от пересечения {pair[0]} + {pair[1]} + {pair[2]}', len(interval_sets[pair[0]] & interval_sets[pair[1]] & interval_sets[pair[2]]))

# 2.1. Определите тип шкалы переменной categoryName.

**CategoryName - номинальная шкала**

Видосы:

1. Гистограмма
2. Боксплот
3. Мода

In [ ]:
# 22 Номинальная статистика

ns = NominalStatistics(df_korean, ['categoryName'])

In [ ]:
# 22 Номинальная статистика вывод
ns.summary()

# EDA

Мода - People & Blogs

In [ ]:
# 22 Гистограмма распределения категорий видео

import plotly.express as px

fig = px.histogram(df_korean, y="categoryName")
fig.show()

In [ ]:
# 23 Типичные видео в рамках категории

df_korean[df_korean['categoryName'] == 'People & Blogs'].sort_values(by='statistics.viewCount', ascending=False).head(5)

## Текстовые характеристики

Предлагается текстовые характеристики описания, которые наиболее часто фигурируют в рамках нашей базы

In [ ]:
# 24 Формирование корпуса текстов
raw_corpora = list(df_korean['snippet.description'].dropna())

print('Длина корпуса предложений:', len(raw_corpora))

for i in range(5):
    print(raw_corpora[i])

In [ ]:
# 25 Удаление символов и паттернов
# email
# пунктуации
# ссылок
# приведение в нижний регистр

import re

def preprocess_text(input_text):

    text = re.sub(r'[A-Za-z0-9]*@[A-Za-z]*\.?[A-Za-z0-9]*', "", input_text)
    text = re.sub(r'http\S+', '', text)
    text = re.sub(r'[^\w\s]', '', text)
    text = " ".join([w.lower() for w in text.split() if len(w) > 2])
    return text
corpora = [preprocess_text(in_text) for in_text in raw_corpora]
print(corpora[0])

In [ ]:
! poetry add nltk

In [ ]:
import nltk

nltk.download('stopwords')

In [ ]:
# 26 Объединение текстов в общий корпус
from nltk.corpus import stopwords

words = []

for sent in corpora:
    for word in sent.split():
        if word not in stopwords.words('russian'):
            words.append(word)

In [ ]:
# 27 Создание частот для анализа
from nltk.probability import FreqDist
fdist = FreqDist(words)
print(fdist)

In [ ]:
# 28 Гистограмма частот
import matplotlib.pyplot as plt
fdist.plot(30,cumulative=False)
plt.show()

In [ ]:
fdist.most_common(20)

In [ ]:
# 29 Добавим stopwords

my_stop = {'видео', 'ссылка', 'инстаграм', 'instagram','iherb', 'подписаться'}

filtered_words = [w for w in words if w not in my_stop]

In [ ]:
! poetry add wordcloud

In [ ]:
# 30 Облако слов

from wordcloud import WordCloud

all_string = ' '.join(filtered_words)

wordcloud = WordCloud(width = 800, height = 800,
                background_color ='white',
                min_font_size = 10).generate(all_string)


# plot the WordCloud image
plt.figure(figsize = (8, 8), facecolor = None)
plt.imshow(wordcloud)
plt.axis("off")
plt.tight_layout(pad = 0)

plt.show()